File: ios/Classes/internal/PowerAuthObjectRegister.swift
/*
 * Copyright 2025 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

internal class PowerAuthObjectRegister {
    
    private let lock = Lock()
    private var register = [String: PowerAuthManagedObject](minimumCapacity: 16)
    private var scheduledCleanup = false
    private var cleanupPeriod = Constants.CLEANUP_PERIOD_DEFAULT
    
    private enum ObjectAction {
        case none
        case use
        case touch
        case remove
    }
    
    // MARK: - Native interface
    
    func registerObject(object: Any, tag: String, policies: [ReleasePolicy]) -> String {
        return lock.synchronized {
            let identifier = generateIdentifier()
            let managedObject = PowerAuthManagedObject(object: object, key: identifier, tag: tag, policies: policies)
            register[identifier] = managedObject
            scheduleClenaup()
            return identifier
        }
    }
    
    func registerObject(object: Any, id: String, tag: String, policies: [ReleasePolicy]) -> Bool {
        return registerObject(id: id, tag: tag, policies: policies) {
            return object
        }
    }
    
    func registerObject(id: String, tag: String, policies: [ReleasePolicy], objectFactory: () -> Any) -> Bool {
        
        return lock.synchronized {
            
            guard self.register[id] == nil else {
                return false
            }
            let object = objectFactory()
            let managedObject = PowerAuthManagedObject(object: object, key: id, tag: tag, policies: policies)
            register[id] = managedObject
            self.scheduleClenaup()
            return true
        }
    }
    
    func useObject<T>(id: String) -> T? {
        return lock.synchronized {
            return findManagedObject(id: id, action: .use)
        }
    }
    
    func findObject<T>(id: String) -> T? {
        return lock.synchronized {
            return findManagedObject(id: id)
        }
    }
    
    func touchObject<T>(id: String) -> T? {
        return lock.synchronized {
            return findManagedObject(id: id, action: .touch)
        }
    }
    
    func containsObject(id: String) -> Bool {
        return lock.synchronized {
            let obj: Any? = self.findManagedObject(id: id)
            return obj != nil
        }
    }
    
    func removeAllObjects(tag: String) {
        lock.synchronized {
            self.findAndRemoveObjects { key, value in
                return tag == value.tag
            }
        }
    }
    
    func removeObject<T>(id: String) -> T? {
        return lock.synchronized {
            return self.findManagedObject(id: id, action: .remove)
        }
    }
    
    func setCleanupPeriod(_ period: Int) {
        lock.synchronized {
            if period >= Constants.CLEANUP_PERIOD_MIN && period <= Constants.CLEANUP_PERIOD_MAX {
                self.cleanupPeriod = period
            } else {
                self.cleanupPeriod = Constants.CLEANUP_PERIOD_DEFAULT
            }
            // Kick the cleanup now, because we don't want to wait for the next
            // tick if period is shorter.
            self.doCleanup()
        }
    }
    
    /// Find object in the object register.
    /// - Parameters:
    ///   - objectId: Generated or application specific object register.
    ///   - expectedClass: Expected class to retrieve. If not provided, then the stored object can be anything.
    ///   - action: Additional operations that will be performed with the object
    /// - Returns: Object retrieved from the register or nil if no such object exist.
    private func findManagedObject<T>(id: String, action: ObjectAction = .none) -> T? {
        
        guard let managedObject = register[id] else {
            return nil
        }
        
        guard let object = managedObject.object as? T else {
            return nil
        }
        
        guard managedObject.isStillValid() else {
            return nil
        }
        
        switch action {
        case .none: break
        case .use: managedObject.setUsed()
        case .touch: managedObject.touch()
        case .remove: register.removeValue(forKey: id)
        }
        return object
    }
    
    /// Schedule an object cleanup job.
    private func scheduleClenaup() {
        guard !scheduledCleanup && register.count > 0 else {
            return
        }
        scheduledCleanup = true
        // Wake-up after cleanup period seconds and do the cleanup.
        DispatchQueue.global().asyncAfter(deadline: .now() + Double(cleanupPeriod) * Double(NSEC_PER_MSEC)) {
            self.lock.synchronized {
                self.doCleanup()
            }
        }
    }
    
    /// Function remove expired or no loner valid objects from the register.
    private func doCleanup() {
        // Clear scheduled flag.
        scheduledCleanup = false
        // Remove all invalid objects
        findAndRemoveObjects { _, obj in
            return obj.isStillValid() == false
        }
        // Schedule cleanup for the next round
        scheduleClenaup()
    }
    
    /// Function find and remove objects from the register. The provided block decide whether object
    /// needs to be removed.
    /// - Parameter block: Block that decide whether object needs to be removed.
    private func findAndRemoveObjects(block: (String, PowerAuthManagedObject) -> Bool) {
        // Find objects that should be removed
        var keysToRemove = [String]()
        for (key, value) in register {
            if block(key, value) {
                keysToRemove.append(key)
            }
        }
        for keyToRemove in keysToRemove {
            register.removeValue(forKey: keyToRemove)
        }
    }
    
    /// Generate a new object identifier.
    /// - Returns: New unique object identifier.
    func generateIdentifier() -> String {
        while(true) {
            let identifier = Self.getRandomString()
            if register[identifier] == nil {
                return identifier
            }
        }
    }
    
//    - (NSArray<NSDictionary*>*) debugDumpObjectsWithTag:(nullable NSString*)tag
//    {
//#if DEBUG
//        return [self synchronized:^{
//            NSMutableArray * content = [NSMutableArray arrayWithCapacity:_register.count];
//            [_register enumerateKeysAndObjectsUsingBlock:^(NSString * key, PowerAuthManagedObject * obj, BOOL * stop) {
//                if (tag && (!obj.tag || ![obj.tag isEqualToString:tag])) {
//                    return;
//                }
//                [content addObject:[obj debugDump]];
//            }];
//            return content;
//        }];
//#else
//        return @[];
//#endif
//    }
    
#if DEBUG
//    - (NSString*) description
//    {
//        return [[self debugDumpObjectsWithTag:nil] description];
//    }
#endif // DEBUG
    
    private static func getRandomString() -> String {
        let count = Int(3 * (3 + arc4random_uniform(6)))
        var data = NSMutableData(length: count)!
        arc4random_buf(data.mutableBytes, data.length)
        return data.base64EncodedString()
    }
    
}

internal class ReleasePolicy: Equatable {
    
    enum RPType: Int {
        case manual = 0
        case afterUse = 1
        case keepAlive = 2
        case expire = 3
    }
    
    private let value: Int
    
    private init(type: RPType, param: Int) {
        self.value = (type.rawValue & 0xF) | (param << 4)
    }
    
    /// Returns the policy type
    func getPolicyType() -> RPType {
        return RPType(rawValue: value & 0xF)!
    }
    
    /// Returns the policy parameter.
    func getPolicyParam() -> Int {
        return value >> 4
    }
    
    /// Creates a new release policy configured for manual release.
    /// This type of policy cannot be combined with other policy types, as the object owner manages the object's lifetime.
    static func manual() -> ReleasePolicy {
        return ReleasePolicy(type: .manual, param: 0)
    }
    
    /// Creates a new release policy configured to release the object after a specified number of uses.
    /// It's recommended to combine this type of policy with `expire()` to ensure the object is always released from memory.
    /// - Parameter count: Maximum number of object uses allowed.
    static func afterUse(_ count: Int) -> ReleasePolicy {
        return ReleasePolicy(type: .afterUse, param: count)
    }
    
    /// Creates a new release policy configured to release the object after a required time of inactivity.
    /// Inactivity means no interaction with the object in the defined time window.
    /// - Parameter timeIntervalMs: Time interval in milliseconds to keep the object alive from the last use attempt.
    static func keepAlive(_ timeIntervalMs: Int) -> ReleasePolicy {
        return ReleasePolicy(type: .keepAlive, param: timeIntervalMs)
    }
    
    /// Creates a new release policy configured to release the object after a required time.
    /// - Parameter timeIntervalMs: Time interval in milliseconds to keep the object alive.
    static func expire(_ timeIntervalMs: Int) -> ReleasePolicy {
        return ReleasePolicy(type: .expire, param: timeIntervalMs)
    }
    
    /// Compares two `ReleasePolicy` objects for equality.
    static func == (lhs: ReleasePolicy, rhs: ReleasePolicy) -> Bool {
        return lhs.value == rhs.value
    }
}

internal class PowerAuthManagedObject {
    
    let object: Any
    let key: String
    let tag: String
    var usageCount: Int = 0
    let createDate: Date
    var lastUseDate: Date
    private let managedByOwner: Bool
    private let policies: [ReleasePolicy]
    
    init(object: Any, key: String, tag: String, policies: [ReleasePolicy]) {
        let now = Date()
        self.object = object
        self.key = key
        self.tag = tag
        self.createDate = now
        self.lastUseDate = now
        self.managedByOwner = policies.contains(.manual())
        self.policies = policies // TODO: filter?
    }
    
    func setUsed() {
        usageCount += 1
        lastUseDate = Date()
    }
    
    func touch() {
        lastUseDate = Date()
    }
    
    
    /// Evaluate whether time interval between now and reference date is greater or equal than time interval in release policy.
    /// - Parameters:
    ///   - refDate: Reference date.
    ///   - rp: Pointer to release policy structure.
    static func isExpired(refDate: Date, rp: ReleasePolicy) -> Bool {
        return -refDate.timeIntervalSinceNow * 1_000 >= Double(rp.getPolicyParam())
    }
    
    func isStillValid() -> Bool {
        // In case that object is manually managed, then don't iterate over the policies.
        guard managedByOwner == false else {
            return true
        }
        
        for policy in policies {
            switch policy.getPolicyType() {
            case .afterUse:
                if usageCount >= policy.getPolicyParam() {
                    return false
                }
                break;
            case .expire:
                if Self.isExpired(refDate: createDate, rp: policy) {
                    return false
                }
            case .keepAlive:
                if Self.isExpired(refDate: lastUseDate, rp: policy) {
                    return false
                }
            case .manual:
                // ignore manual (also should never
                break
            }
        }
        return true
    }
    
    func debugDump() -> [String: Any?] {
#if DEBUG
        // Iterate over policies
        var printLastUseDate = false
        var printUsageCount = false
        var policies = [String]()
        if managedByOwner {
            policies.append("MANUAL")
        } else {
            for policy in self.policies {
                switch policy.getPolicyType() {
                case .afterUse:
                    policies.append("AFTER_USE(\(usageCount)/\(policy.getPolicyParam()))")
                    printUsageCount = true
                case .keepAlive:
                    policies.append("KEEP_ALIVE(\(policy.getPolicyParam()))")
                    printLastUseDate = true
                case .expire:
                    policies.append("EXPIRE(\(policy.getPolicyParam()))")
                case .manual:
                    break
                }
            }
        }
        return [
            "id": key,
            "class": "\(object.self)",
            "tag": tag,
            "createDate": createDate.timeIntervalSince1970,
            "lastUseDate": printLastUseDate ? lastUseDate.timeIntervalSince1970 : nil,
            "usageCount": printUsageCount ? usageCount : nil,
            "policies": policies,
            "isValid": isStillValid()
        ]
#else
        return [:]
#endif // DEBUG
    }
}

File: android/src/main/kotlin/com/wultra/android/powerauth/flutter/PowerAuthObjectRegister.kt
/*
 * Copyright 2025 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wultra.android.powerauth.flutter

import android.os.SystemClock
import android.util.Base64
import com.wultra.android.powerauth.flutter.Constants.CLEANUP_PERIOD_DEFAULT
import com.wultra.android.powerauth.flutter.Constants.CLEANUP_PERIOD_MAX
import com.wultra.android.powerauth.flutter.Constants.CLEANUP_PERIOD_MIN
import com.wultra.android.powerauth.flutter.Constants.CLEANUP_REMOVE_DELAY
import java.util.Random
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.locks.ReentrantLock
import kotlin.collections.ArrayList
import kotlin.concurrent.withLock
import io.getlime.security.powerauth.core.Password
import java.nio.charset.StandardCharsets

/**
 * Object register that allows exposing native objects.
 * The object is identified by a unique identifier created at the time of registration
 * or by an application-provided identifier.
 */
class PowerAuthObjectRegister private constructor(private val isDebug: Boolean) {

    companion object {
        @Volatile
        private var INSTANCE: PowerAuthObjectRegister? = null

        @Volatile
        private var attachmentCount = 0

        private const val OPT_NONE = 0
        private const val OPT_SET_USE = 1
        private const val OPT_TOUCH = 2
        private const val OPT_REMOVE = 3

        @JvmStatic
        fun getInstance(isDebug: Boolean = false): PowerAuthObjectRegister {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: PowerAuthObjectRegister(isDebug).also { INSTANCE = it }
            }
        }

        @JvmStatic
        internal fun onPluginAttached() {
            synchronized(this) {
                attachmentCount++
            }
        }

        @JvmStatic
        internal fun onPluginDetached() {
            synchronized(this) {
                attachmentCount = maxOf(0, attachmentCount - 1)

                // If no more plugins / engines are attached, we can clean up the register.
                // TODO: do we want to keep the singleton instance for potential re-init?
                if (attachmentCount == 0) {
                    INSTANCE?.removeAllObjects()
                }
            }
        }


        // For testing purposes - TODO: remove before release
        @JvmStatic
        internal fun resetForTesting() {
            synchronized(this) {
                INSTANCE?.removeAllObjects()
                INSTANCE = null
                attachmentCount = 0
            }
        }

        @JvmStatic
        internal fun getAttachmentCount(): Int {
            synchronized(this) {
                return attachmentCount
            }
        }
    }

    private val lock = ReentrantLock(false)
    private val managedObjects = mutableMapOf<String, ManagedObjectHolder>()
    private val randomGenerator: Random = Random()
    private var cleanupTimer: Timer? = null
    private var cleanupPeriodMs: Long = CLEANUP_PERIOD_DEFAULT.toLong()

    /**
     * Factory that creates an object on demand.
     */
    fun interface ObjectFactory<T : Any> {
        @Throws(Throwable::class)
        fun create(): IManagedObject<T>
    }

    private class ManagedObjectHolder(
        val obj: IManagedObject<out Any>,
        val tag: String?,
        policies: List<ReleasePolicy>,
        val creationTime: Long = SystemClock.elapsedRealtime(),
        var lastUseTime: Long = creationTime,
        var useCount: Int = 0,
        var removeOrderTime: Long = 0
    ) {
        // TODO: improve
        val policies = if (policies.contains(ReleasePolicy.manual())) null else policies

        fun setUsed() {
            lastUseTime = SystemClock.elapsedRealtime()
            useCount++
        }

        fun touch() {
            lastUseTime = SystemClock.elapsedRealtime()
        }

        /**
         * Marks object as removed and return information whether object can be removed immediately.
         */
        fun setRemoved(): Boolean {
            removeOrderTime = SystemClock.elapsedRealtime()
            // If policies is null then the object is manually managed, so it should be removed immediately.
            return policies == null
        }

        // TODO: refactor the null-defaulting into this if it makes sense
        val isManuallyManaged: Boolean
            get() {
                return (policies?.contains(ReleasePolicy.manual()) == true)
            }

        /**
         * Determine whether this native object is still valid.
         */
        val isStillValid: Boolean
            get() {
                if (removeOrderTime != 0L) {
                    return false
                }

                if (policies == null) {
                    return true
                }

                val currentTime = SystemClock.elapsedRealtime()
                policies.forEach { rp ->
                    val param = rp.getPolicyParam()

                    when (rp.getPolicyType()) {
                        ReleasePolicy.TYPE_AFTER_USE -> if (useCount >= param) return false
                        ReleasePolicy.TYPE_KEEP_ALIVE -> if (currentTime - lastUseTime >= param.toLong()) return false
                        ReleasePolicy.TYPE_EXPIRE -> if (currentTime - creationTime >= param.toLong()) return false
                    }
                }

                return true
            }

        /**
         * Determine whether object can be removed from the register.
         */
        val isReadyForRemove: Boolean
            get() {
                if (isStillValid) {
                    return false
                }

                // If removeOrderTime is 0, then object was not explicitly removed.
                // This means that object is expired or was used for a limited number of times.
                // On other side, if time is specified then the object was explicitly removed, so
                // it should be ready for remove after a short delay period.
                val readyNow =
                    removeOrderTime == 0L || (SystemClock.elapsedRealtime() - removeOrderTime >= CLEANUP_REMOVE_DELAY.toLong())

                if (readyNow) {
                    obj.cleanup()
                }

                return readyNow
            }

        fun debugDump(): Map<String, Any?> {
            return mapOf(
                "class" to obj.managedInstance()::class.java.simpleName,
                "isValid" to isStillValid,
                "tag" to tag,
                "createDate" to creationTime,
                "lastUseDate" to if (lastUseTime != creationTime) lastUseTime else null,
                "usageCount" to useCount,
                "policies" to (policies?.map { it.toString() } ?: emptyList())
            )
        }
    }

    /**
     * Registers an object and returns its unique identifier.
     */
    fun <T : Any> registerObject(
        objectWrapper: IManagedObject<T>,
        tag: String?,
        releasePolicies: List<ReleasePolicy>
    ): String = lock.withLock {
        val id = generateId()

        managedObjects[id] = ManagedObjectHolder(objectWrapper, tag, releasePolicies)
        scheduleCleanupJob()

        return@withLock id
    }

    /**
     * Registers an object with an application-provided identifier.
     */
    fun <T : Any> registerObjectWithId(
        id: String,
        objectWrapper: IManagedObject<T>,
        tag: String?,
        releasePolicies: List<ReleasePolicy>
    ): Boolean = lock.withLock {
        if (!isValidObjectId(id) || managedObjects.containsKey(id)) {
            return@withLock false
        }

        managedObjects[id] = ManagedObjectHolder(objectWrapper, tag, releasePolicies)
        scheduleCleanupJob()

        return@withLock true
    }

    /**
     * Registers an object provided by a factory with an application-provided identifier.
     */
    @Throws(Throwable::class)
    fun <T : Any> registerObjectWithId(
        id: String,
        tag: String?,
        releasePolicies: List<ReleasePolicy>,
        factory: ObjectFactory<T>
    ): Boolean = lock.withLock {
        if (!isValidObjectId(id) || managedObjects.containsKey(id)) {
            return@withLock false
        }

        val objectWrapper = factory.create()

        managedObjects[id] = ManagedObjectHolder(objectWrapper, tag, releasePolicies)
        scheduleCleanupJob()

        return@withLock true
    }

    /**
     * Find object with given identifier and do an additional operation with the object.
     * @param id Object identifier.
     * @param expectedClass Expected class, or null if any object can be returned (in case of remove)
     * @param options Additional operation that should be performed with the object's entry. Use `OPT_*` constants.
     * @param <T> Expected object's type.
     * @return instance of object with given identifier or null if no such object exists in register.
    </T> */
    private fun <T : Any> findAndProcessObject(
        id: String?,
        expectedClass: Class<T>?,
        options: Int
    ): T? {
        val objectId = id ?: return null
        val holder = managedObjects[objectId] ?: return null

        val instance = holder.obj.managedInstance()

        if (!holder.isStillValid ||
            (expectedClass != null && !expectedClass.isInstance(instance))
        ) return null

        when (options) {
            OPT_SET_USE -> holder.setUsed()
            OPT_TOUCH -> holder.touch()
            OPT_REMOVE -> if (holder.setRemoved()) {
                holder.obj.cleanup()
                managedObjects.remove(id)
            }
        }

        @Suppress("UNCHECKED_CAST")
        return instance as T
    }

    fun <T : Any> findObject(id: String, type: Class<T>): T? = lock.withLock {
        return findAndProcessObject(id, type, OPT_NONE)
    }

    fun <T : Any> useObject(id: String, type: Class<T>): T? = lock.withLock {
        return findAndProcessObject(id, type, OPT_SET_USE)
    }

    fun <T : Any> touchObject(id: String, type: Class<T>): T? = lock.withLock {
        return findAndProcessObject(id, type, OPT_TOUCH)
    }

    fun <T : Any> removeObject(id: String, type: Class<T>): T? = lock.withLock {
        return@withLock findAndProcessObject(id, type, OPT_REMOVE)
    }

    /**
     * Removes all objects associated with a specific tag.
     * If tag is null, removes all objects that are not manually managed.
     */
    fun removeAllObjectsWithTag(tag: String?) = lock.withLock {
        val iterator = managedObjects.entries.iterator()
        var changed = false
        while (iterator.hasNext()) {
            val entry = iterator.next()
            val holder = entry.value

            // TODO: implement proper filtering!
            if (tag == null || holder.tag == tag) {
                if (holder.setRemoved()) {
                    holder.obj.cleanup()
                    iterator.remove()

                    changed = true
                } else if (holder.isReadyForRemove) {
                    iterator.remove()
                    changed = true
                } else {
                    // marked for later cleanup, ensure a job is scheduled
                    changed = true
                }
            }
        }

        if (changed) scheduleCleanupJob()
    }

    /**
     * Removes all objects from the register, regardless of policy.
     */
    internal fun removeAllObjects() = lock.withLock {
        managedObjects.values.forEach { it.obj.cleanup() }
        managedObjects.clear()
        stopCleanupJob()
    }

    fun isValidObjectId(id: String?): Boolean {
        return !id.isNullOrEmpty()
    }

    private fun setCleanupPeriod(periodMs: Long) = lock.withLock {
        cleanupPeriodMs = if (periodMs in CLEANUP_PERIOD_MIN..CLEANUP_PERIOD_MAX) {
            periodMs
        } else {
            CLEANUP_PERIOD_DEFAULT.toLong()
        }

        if (cleanupTimer != null) {
            stopCleanupJob()
            startCleanupJob()
        }
    }

    private fun generateId(): String {
        val numberOfBytes = 3 * (3 + randomGenerator.nextInt(6))
        val randomBytes = ByteArray(numberOfBytes)

        while (true) {
            randomGenerator.nextBytes(randomBytes)
            val id = Base64.encodeToString(randomBytes, Base64.NO_WRAP)

            if (!managedObjects.containsKey(id)) {
                return id
            }
        }
    }

    /** Schedule an object cleanup job. */
    private fun scheduleCleanupJob() = lock.withLock {
        if (managedObjects.isNotEmpty()) {
            if (cleanupTimer == null) {
                cleanupTimer = Timer("PowerAuthObjectRegisterTimer")
                startCleanupJob()
            }
        } else {
            stopCleanupJob()
        }
    }

    private fun startCleanupJob() {
        if (cleanupTimer == null) return

        try {
            cleanupTimer?.schedule(object : TimerTask() {
                override fun run() {
                    lock.withLock { performCleanup() }
                }
            }, cleanupPeriodMs, cleanupPeriodMs)
        } catch (_: IllegalStateException) {
        }
    }

    private fun stopCleanupJob() {
        cleanupTimer?.cancel()
        cleanupTimer?.purge()
        cleanupTimer = null
    }

    /** Function remove expired or no longer valid objects from the register. */
    private fun performCleanup() = lock.withLock {
        val idsToRemove = ArrayList<String>()
        managedObjects.forEach { (id, holder) ->
            if (holder.isReadyForRemove) {
                idsToRemove.add(id)
            }
        }

        if (idsToRemove.isNotEmpty()) {
            idsToRemove.forEach { managedObjects.remove(it) }
        }

        scheduleCleanupJob()
    }

    // Called by PowerAuthPlugin when detaching from Flutter
    fun invalidate() {
        lock.withLock {
            removeAllObjects()
        }
    }

    fun debugDumpObjectsWithTag(tag: String?): List<Map<String, Any?>> {
        if (!isDebug) {
            return emptyList()
        }

        return lock.withLock {
            val result = mutableListOf<Map<String, Any?>>()
            for ((key, value) in managedObjects) {
                if (tag == null || tag == value.tag) {
                    val dump = value.debugDump().toMutableMap()
                    dump["id"] = key
                    result.add(dump)
                }
            }

            return@withLock result
        }
    }

    fun debugCommand(command: String, data: Map<String, Any>): Any? {
        if (!isDebug) {
            return null
        }

        val objectId = data["objectId"] as? String
        when (command) {
            "create" -> {
                val objectType = data["objectType"] as? String
                val objectTag = data["objectTag"] as? String
                val releasePolicyDescription = data["releasePolicy"] as? List<String>

                val policies = mutableListOf<ReleasePolicy>()

                if (releasePolicyDescription != null) {
                    for (policy in releasePolicyDescription) {
                        val components = policy.split(" ".toRegex()).toTypedArray()
                        val param = if (components.size > 1) components[1].toInt() else 1

                        when {
                            policy.startsWith("manual") -> policies.add(ReleasePolicy.manual())
                            policy.startsWith("afterUse") -> policies.add(
                                ReleasePolicy.afterUse(
                                    param
                                )
                            )

                            policy.startsWith("keepAlive") -> policies.add(
                                ReleasePolicy.keepAlive(
                                    param
                                )
                            )

                            policy.startsWith("expire") -> policies.add(ReleasePolicy.expire(param))
                        }
                    }
                }

                if (policies.isNotEmpty()) {
                    val newObject: IManagedObject<out Any>? = when (objectType) {
                        "data" -> {
                            val td = "TEST-DATA".toByteArray(StandardCharsets.UTF_8)
                            ManagedAny.wrap(td, null)
                        }

                        "secureData" -> {
                            val td = "SECURE-DATA".toByteArray(StandardCharsets.UTF_8)
                            ManagedAny.wrap(td)
                        }

                        "number" -> ManagedAny.wrap(42)
                        "password" -> ManagedAny.wrap(
                            Password(),
                            cleanupAction = { password -> password.destroy() })

                        else -> null
                    }
                    if (newObject != null) {
                        return registerObject(newObject, objectTag, policies)
                    }
                }

                return null
            }

            "release" -> {
                val objectIdNonNull = objectId ?: throw WrapperException(
                    Errors.EC_WRONG_PARAMETER,
                    "Missing objectId"
                )
                val objectType = data["objectType"] as? String
                val clazz = getClassForObjectType(objectType)
                    ?: throw WrapperException(Errors.EC_WRONG_PARAMETER, "Unknown objectType")
                val removedInstance = removeObject(objectIdNonNull, clazz)

                return removedInstance != null
            }

            "releaseAll" -> {
                val tag = data["objectTag"] as? String
                removeAllObjectsWithTag(tag)

                return null
            }

            "use" -> {
                val objectIdNonNull = objectId ?: throw WrapperException(
                    Errors.EC_WRONG_PARAMETER,
                    "Missing objectId"
                )
                val objectType = data["objectType"] as? String
                val clazz = getClassForObjectType(objectType)
                    ?: throw WrapperException(Errors.EC_WRONG_PARAMETER, "Unknown objectType")

                return useObject(objectIdNonNull, clazz) != null
            }

            "find" -> {
                val objectIdNonNull = objectId ?: throw WrapperException(
                    Errors.EC_WRONG_PARAMETER,
                    "Missing objectId"
                )
                val objectType = data["objectType"] as? String
                val clazz = getClassForObjectType(objectType)
                    ?: throw WrapperException(Errors.EC_WRONG_PARAMETER, "Unknown objectType")

                return findObject(objectIdNonNull, clazz) != null
            }

            "touch" -> {
                val objectIdNonNull = objectId ?: throw WrapperException(
                    Errors.EC_WRONG_PARAMETER,
                    "Missing objectId"
                )
                val objectType = data["objectType"] as? String
                val clazz = getClassForObjectType(objectType)
                    ?: throw WrapperException(Errors.EC_WRONG_PARAMETER, "Unknown objectType")

                return touchObject(objectIdNonNull, clazz) != null
            }

            "setPeriod" -> {
                val period = data["cleanupPeriod"] as? Int
                if (period != null) {
                    setCleanupPeriod(period.toLong())
                }

                return null
            }

            else -> throw WrapperException(
                Errors.EC_WRONG_PARAMETER,
                "Unsupported debug command: $command"
            )
        }
    }

    private fun getClassForObjectType(objectType: String?): Class<out Any>? {
        return when (objectType) {
            "data", "secureData" -> ByteArray::class.java
            "number" -> Number::class.java
            "password" -> Password::class.java
            else -> null
        }
    }
}
